### PR TITLE
Add draft libc post

### PR DESCRIPTION
**Pull request by franc0is**

### CODE REVIEW

File: _drafts/zero-to-main-libc.md [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+SRCS = \
+	startup_samd21.c \
+	$(PROJECT).c
+
+include ../common-standalone.mk
+```
+
+It is very easy to add a dependency on the C standard library without meaning to, as GCC will sometimes use standard C functions implicitly. For example, consider this code used to zero-initialize a struct:
+
+```c
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Ooo, awesome example!

Nit: I'd probably use  a slightly bigger buffer in the example, below a certain size the memset will usually be unrolled ...

```
extern void noop(void *p);
__attribute__((noinline))
void example_memset(void) {
    int b[3] = { 0 };
    noop(b);
}

// int b[3] = { 0 };
b4:       b51f            push    {r0, r1, r2, r3, r4, lr}
b6:       2300            movs    r3, #0
b8:       a801            add     r0, sp, #4
ba:       e9cd 3301       strd    r3, r3, [sp, #4]
be:       9303            str     r3, [sp, #12]
c0:       f000 f863       bl      18a <noop>
c4:       b005            add     sp, #20
c6:       f85d fb04       ldr.w   pc, [sp], #4


// int b[50] = { 0 };
b4:       b500            push    {lr}
b6:       b0b3            sub     sp, #204        ; 0xcc
b8:       22c8            movs    r2, #200        ; 0xc8
ba:       2100            movs    r1, #0
bc:       4668            mov     r0, sp
be:       f001 fe63       bl      1d88 <memset>
c2:       4668            mov     r0, sp
c4:       f000 f863       bl      18e <noop>
c8:       b033            add     sp, #204        ; 0xcc
ca:       f85d fb04       ldr.w   pc, [sp], #4
```

Aside: I thought you could convince the compiler to not use any libc calls with some combination of fno-builtin, -ffreestanding, etc but I couldn't actually get the b[50] example to compile. Always hit:

```
/Users/chrisc/arm_8_2019q3/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: /Users/chrisc/dev/interrupt/example/freertos-example/build/main.o: in function `example_memset':
/Users/chrisc/dev/interrupt/example/freertos-example/main.c:57: undefined reference to `memset'
collect2: error: ld returned 1 exit status
```

File: _drafts/zero-to-main-libc.md [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+isattyr.c:(.text._isatty_r+0xc): undefined reference to `_isatty'
+collect2: error: ld returned 1 exit status
+```
+
+Specifically, the compiler is asking for `_fstat`, `_read`, `_lseek`, `_close`, `_write`, and `_sbrk`.
+
+The [newlib documentation](https://sourceware.org/newlib/libc.html#Syscalls) calls these functions “system calls”. In short, they are the handful of things newlib expects the underlying “operating system”. The complete list of them is provided below:
+
+```
+_exit, close, environ, execve, fork, fstat, getpid, isatty, kill, link, lseek, open, read, sbrk, stat, times, unlink, wait, write
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Maybe: Might format these so you don't have to horizontally scroll to see them all

File: _drafts/zero-to-main-libc.md [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+This is where things get interesting! `write` writes up to `count` bytes from the buffer starting at `buf` to the file referred to by the file descriptor `fd`.
+
+Functions like `printf` rely on `write` to write bytes to `STDOUT`. In our case, we will want those bytes to be written to serial instead.
+
+On the SAMD21 chip we are using, writing bytes to serial is done using the `usart_serial_putchar` function. We can use it to implement `write`:
+
+```c
+static struct usart_module stdio_uart_module;
+
+int _write (int fd, char *buf, int count) {
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Not necessary but you could also check that fd is one of the [std streams](https://github.com/bminor/newlib/blob/master/newlib/libc/stdio/findfp.c#L90-L118). Should always be 0,1,2 for [posix](https://pubs.opengroup.org/onlinepubs/9699919799/functions/stdin.html)

File: _drafts/zero-to-main-libc.md [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+
+`sbrk` increases the program’s data space by `increment` bytes. In other words, it increases the size of the heap.
+
+What does `printf` have to do with the heap, you will justly ask? It turns out that newlib’s `printf` implementations allocates data on the heap and depends on a working `malloc` implementation. 
+
+The source for `printf` is hard to follow, but you will find that indeed [it calls malloc](https://github.com/bminor/newlib/blob/6497fdfaf41d47e835fdefc78ecb0a934875d7cf/newlib/libc/stdio/vfprintf.c#L226)!
+
+Here’s a simple implementation of `sbrk`:
+
+```
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: `c`

File: _drafts/zero-to-main-libc.md [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+caddr_t _sbrk(int incr) {
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Some kind of check we haven't hit heap end?

nit: I think newlib just used void * instead of caddr_t:

```
../include/sys/unistd.h
275:void *  _sbrk (ptrdiff_t __incr);
```

File: _drafts/zero-to-main-libc.md [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+  static unsigned char *heap = HEAP_START;
+  unsigned char *prev_heap = heap;
+  heap += incr;
+  return (caddr_t) prev_heap;
+}
+```
+
+More often than not, we want the heap to use all the RAM not used by anything else. We therefore set `HEAP_START` to the first address not spoken for in our linker script. In our [previous post](https://interrupt.memfault.com/blog/how-to-write-linker-scripts-for-firmware#complete-linker-script), we had added the `_end` variable in our linker script to that end.
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: OT: What I don't like about this approach is as you add more data to your code, the heap silently shrinks on you increasing the chances of OOMs. Personally a fan of using a hardcoded static buffer in the C code itself

File: _drafts/zero-to-main-libc.md [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+
+We replace `HEAP_START` with `_end` and get:
+
+```
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: c

File: _drafts/zero-to-main-libc.md [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+  }
+  prev_heap = heap;
+
+  heap += incr;
+
+  return (caddr_t) prev_heap;
+}
+```
+
+### Initializing State with Constructors & Destructors
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Just noting, still a TODO

File: _drafts/zero-to-main-libc.md [hunk 9]
```diff
@@ -0,0 +1,385 @@
+---
+title: "From zero to main(): Bootstrapping libc"
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Would probably work "newlib" somewhere into the title and/or description.

Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: :+1:

File: _drafts/zero-to-main-libc.md [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+```c
+#include <stdio.h>
+
+int main() {
+  printf("Hello, World\n");
+  while (1) {}
+}
+```
+
+Compiling this using our makefile and linker script from previous posts, we hit the following error:
```
[the remark this answers is not in the document]
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+<!-- excerpt start -->
+In this post, we will add RedHat’s Newlib to our firmware and highlight some of its features. We will implement syscalls, learn about constructors, and finally print out “Hello, World”!
+<!-- excerpt end -->
+
+## Setup
+As we did in our previous posts, we are using Adafruit’s Metro M0 development board to run our examples. We use a cheap CMSIS-DAP adapter and openOCD to program it.
+
+You can find a step by step guide [in our previous post](https://interrupt.memfault.com/blog/getting-started-with-ibdap-and-atsamd21g18).
+
+As with previous examples, we start with our “minimal” example which you can find [on GitHub](https://github.com/memfault/zero-to-main/tree/master/minimal). I’ve reproduce the source code for `main.c` below:
```
[the remark this answers is not in the document]
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 1 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 2 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 3 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: My initial implementation did that, but it warrants more discussion which isn't all that valuable.

File: _drafts/zero-to-main-libc.md (hunk 4 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 5 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 6 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: You have a point. That would also be simpler w.r.t. the linker script (no need for the variable). At the same time, this is good enough and I'd rather not rewrite it.

File: _drafts/zero-to-main-libc.md (hunk 7 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 8 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md [hunk 12]
```diff
@@ -0,0 +1,721 @@
+---
+title: "From zero to main(): Bootstrapping libc with Newlib"
+description:
+  "TODO"
+author: francois
+tags: [zero-to-main]
+---
+
+In this series of post, we’ve worked methodically to demystify what happens to
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: In the 'Zero to Main()' series of post[s]

Maybe add a link too.

File: _drafts/zero-to-main-libc.md [hunk 13]
```diff
@@ -0,0 +1,721 @@
...
+firmware before the main() function is called. So far, we bootstrapped a C
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: maybe: within firmware

File: _drafts/zero-to-main-libc.md [hunk 14]
```diff
@@ -0,0 +1,721 @@
...
+environment, wrote a linker script from scratch, and implemented our own
+bootloader.
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: Tense nit, but if you are going to start with `we have worked`, you might want to change your later tense to `we have bootstrapped, written a linker script,`

File: _drafts/zero-to-main-libc.md [hunk 15]
```diff
@@ -0,0 +1,721 @@
...
+
+```c
+#include <stdio.h>
+
+int main() {
+  printf("Hello, World\n");
+  while (1) {}
+}
+```
+Compiling this using our makefile and linker script from [previous](% post_url
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: Link is broken 

`[post]({% post_url 2019-09-17-continuous-integration-for-firmware %})`

File: _drafts/zero-to-main-libc.md [hunk 16]
```diff
@@ -0,0 +1,721 @@
...
+2019-05-14-zero-to-main-1 %) [posts](% post_url
+2019-06-25-how-to-write-linker-scripts-for-firmware %), we hit
+the following error:
+
+```shell
+$ make
+...
+Linking build/minimal.elf
+/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: build/objs/a/b/c/minimal.o: in function `main':
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: The full error is nice, but I'd make it shorter or wrap. No one really cares about the exact content of the error, but it's nice to have context.

File: _drafts/zero-to-main-libc.md [hunk 17]
```diff
@@ -0,0 +1,721 @@
...
+/minimal/minimal.c:4: undefined reference to `printf'
+collect2: error: ld returned 1 exit status
+make: *** [build/minimal.elf] Error 1
+```
+
+Undefined reference to `printf`! How could this be? Our firmware’s C environment
+is still missing a key component: a working C standard library.  This means that
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: I'd love to have `working C standard library` bolded, but that's me

File: _drafts/zero-to-main-libc.md [hunk 18]
```diff
@@ -0,0 +1,721 @@
...
+commonly used functions such as `printf`, `memcpy`, or `strncpy` are all out of
+reach of our program so far.
+
+In firmware-land, nothing comes free with the system: just like we had to
+explicitly zero out the `bss` region to initialize some of our static variables,
+we’ll have to port a `printf` implementation alongside a C standard library if
+we want to use them.
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: use [it]

File: _drafts/zero-to-main-libc.md [hunk 19]
```diff
@@ -0,0 +1,721 @@
...
+  }
+}
+``` 
+
+## Implementing Newlib
+### Why Newlib?
+
+There are several implementations of the C Standard Library, starting with the
+venerable `glibc` found on most GNU/Linux systems. Alternative implementations
+include Musl libc, Bionic libc, ucLibc, and dietlibc.
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: Would love some juicy links to these.

File: _drafts/zero-to-main-libc.md [hunk 20]
```diff
@@ -0,0 +1,721 @@
...
+   1292       0    8192    9484    250c build/minimal.elf
+```
+
+So here we are, using Newlib, and we did not have to do anything. Could it
+really be this simple?
+
+> Note: the variant of Newlib bundled with `arm-none-eabi-gcc` is not compiled
+> with `-g`, which can make debugging difficult. For that reason, you may chose
+> to replace it with your own build of Newlib. You can read more about that
+> process in the "Implementing our own C standard library" section of this
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: process in the `[Implementing our own C standard library](#implementing-our-own-c-standard-library)` section of this

Should provide a link

File: _drafts/zero-to-main-libc.md [hunk 21]
```diff
@@ -0,0 +1,721 @@
...
+If we want `printf` to work anywhere in our `main` function, then `serial_init`
+must be run **before** main. What runs before main? We know from our previous
+post that it is the `Reset_Handler`. A simple solution might therefore be:
+
+```c
+void Reset_Handler(void)
+{
+        /* ... */
+        /* Hardware Initialization */
+        serial_init();
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: nit: better indentation?

File: _drafts/zero-to-main-libc.md [hunk 22]
```diff
@@ -0,0 +1,721 @@
...
+
+#### Locking shared memory
+
+Some standard library functions depend on global memory which would not make
+sense to hold in the `_reent` struct. This is especially important when using
+`malloc` to allocate memory out of the heap. If mutliple threads try modifying
+the heap at the same time, they risk corrupting it.
+
+To allow multiple threads to call `malloc`, Newlib provides the `__malloc_lock`
+and `__malloc_unlock` APIs[^1]. A good implementation of these APIs would lock
```
Comment (tyhoff) [inline _drafts/zero-to-main-libc.md]: link not working

File: _drafts/zero-to-main-libc.md (hunk 12 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 14 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 15 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 16 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 17 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 🙅‍♂

File: _drafts/zero-to-main-libc.md (hunk 18 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 19 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 20 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 21 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 22 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md [hunk 23]
```diff
@@ -0,0 +1,726 @@
...
+arm-none-eabi-objcopy build/minimal.elf build/minimal.bin -O binary
+arm-none-eabi-size build/minimal.elf
+   text    data     bss     dec     hex filename
+   1292       0    8192    9484    250c build/minimal.elf
+```
+
+So here we are, using Newlib, and we did not have to do anything. Could it
+really be this simple?
+
+> Note: the variant of Newlib bundled with `arm-none-eabi-gcc` is not compiled
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: ⭐️

File: _drafts/zero-to-main-libc.md [hunk 24]
```diff
@@ -0,0 +1,726 @@
...
+  usart_serial_init(&stdio_uart_module, SERCOM5, &usart_conf);
+  usart_enable(&stdio_uart_module);
+}
+```
+
+But how do these constructors get invoked? We know that in firmware, we do not
+get anything for free. This is where newlib comes in.
+
+By default, GCC will put every constructor into an array in their own section of
+flash. Newlib then offers a function, `__libc_init_array` which iterates over
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: neat! I haven't actually leveraged this before

File: _drafts/zero-to-main-libc.md [hunk 25]
```diff
@@ -0,0 +1,726 @@
...
+### Replacing a function
+
+Because Newlib is a static library with a separate object file for every
+function, all you need to do to replace a function is define it in your program.
+The linker won’t go looking for it in static libraries if it finds it in your
+code.
+
+For example, we may want to replace Newlib’s `printf` implementation, either
+because it is too large or because it depends on dynamic memory management.
+Using Marco Paland’s [excellent alternative](https://github.com/mpaland/printf)
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: <3

File: _drafts/zero-to-main-libc.md [hunk 26]
```diff
@@ -0,0 +1,726 @@
...
+Once we have copied the static lib (`.a`) for our selected libc, we disable
+Newlib with `-nostdlib` and explicitly link in our substitute library. You can
+find the resulting Makefile below:
+
+```makefile
+PROJECT := with-libc
+BUILD_DIR ?= build
+
+CFLAGS += -nostdlib
+LDFLAGS += -L../lib/embeddedartistry_libc/libc.a -lc
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: nit: I don't think the libc at the end of the -L is needed ...

```
  -L DIRECTORY, --library-path DIRECTORY
                              Add DIRECTORY to library search path
```

File: _drafts/zero-to-main-libc.md [hunk 27]
```diff
@@ -0,0 +1,726 @@
...
+	startup_samd21.c \
+	$(PROJECT).c
+
+include ../common-standalone.mk
+``` 
+
+> Note that the `__libc_init_array` functionality is not found in every standard
+> C library. You will either need to avoid using it, or bring in Newlib’s
+> implementation.
+
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Could drop our typical conclusion blurb here if you want

File: _drafts/zero-to-main-libc.md [hunk 28]
```diff
@@ -0,0 +1,726 @@
...
+commonly used functions such as `printf`, `memcpy`, or `strncpy` are all out of
+reach of our program so far.
+
+In firmware-land, nothing comes free with the system: just like we had to
+explicitly zero out the `bss` region to initialize some of our static variables,
+we’ll have to port a `printf` implementation alongside a C standard library if
+we want to use it.
+
+<!-- excerpt start -->
+In this post, we will add RedHat’s Newlib to our firmware and highlight some of
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Maybe: a reference link here too

File: _drafts/zero-to-main-libc.md [hunk 29]
```diff
@@ -0,0 +1,726 @@
+---
+title: "From zero to main(): Bootstrapping libc with Newlib"
+description:
```
Comment (chrisc11) [inline _drafts/zero-to-main-libc.md]: Just a reminder, still flagged as a TODO

File: _drafts/zero-to-main-libc.md (hunk 26 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 27 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍

File: _drafts/zero-to-main-libc.md (hunk 29 above)
Comment (franc0is) [reply _drafts/zero-to-main-libc.md]: 👍
